state/raftlease/target.go

```diff
@@ -174,33 +176,110 @@ func (t *notifyTarget) Claimed(key lease.Key, holder string) error {
+	if len(expiries) == 0 {
+		return nil
+	}
+
 	coll, closer := t.mongo.GetCollection(t.collection)
 	defer closer()
 
-	docId := leaseHolderDocId(key.Namespace, key.ModelUUID, key.Lease)
-	t.logger.Infof("expiring lease %q", docId)
+	// Cache all the document idents up front, incase we have to retry the
+	// transaction again. Also this serves as a de-duping process.
+	leaseDocs := make(map[string]leaseDoc)
+	for _, expired := range expiries {
+		key := expired.Key
+		docId := leaseHolderDocId(key.Namespace, key.ModelUUID, key.Lease)
+
+		// We have an existing lease.Key already in the documents that doesn't
+		// have the same holder.
+		if doc, ok := leaseDocs[docId]; ok && doc.Holder != expired.Holder {
+			t.logger.Warningf("multiple expiries with the same key %q, but different holders %q and %q", key, doc.Holder, expired.Holder)
+			continue
+		}
+
+		leaseDocs[docId] = leaseDoc{
+			Key:    key,
+			DocId:  docId,
+			Holder: expired.Holder,
+		}
+	}
+	docIds := set.NewStrings()
+	for _, leaseDoc := range leaseDocs {
+		docIds.Add(leaseDoc.DocId)
+	}
+	sortedDocIds := docIds.SortedValues()
+	t.logger.Infof("expiring leases %v", sortedDocIds)
 
 	err := t.mongo.RunTransaction(func(_ int) ([]txn.Op, error) {
-		existingDoc, err := getRecord(coll, docId)
-		if err == mgo.ErrNotFound {
+		// Bulk get the records, to prevent potato programming.
+		existingDocs, err := getRecords(coll, sortedDocIds)
+		if errors.Cause(err) == mgo.ErrNotFound {
+			t.logger.Warningf("unable to expire %v, no documents found", sortedDocIds)
 			return nil, jujutxn.ErrNoOperations
 		}
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
-		return []txn.Op{{
-			C:  t.collection,
-			Id: docId,
-			Assert: bson.M{
-				fieldHolder: existingDoc.Holder,
-			},
-			Remove: true,
-		}}, nil
+
+		// Ensure that we have all the docIds represented in the existingDocs
+		// and that we haven't missed any.
+		existingDocsSet := set.NewStrings()
+		for _, doc := range existingDocs {
+			existingDocsSet.Add(doc.Id)
+		}
+		if missing := docIds.Difference(existingDocsSet); len(missing) > 0 {
+			// Not all documents are represented by the returned records query
+			// from mongo. Report a warning, but continue and expire as much as
+			// possible.
+			t.logger.Warningf("missing expiry documents: %v, unable to do a complete expiry", missing)
+		}
+
+		ops := make([]txn.Op, 0)
+		for _, doc := range existingDocs {
+			leaseDoc, ok := leaseDocs[doc.Id]
+			if !ok {
+				// This should never happen, we have an existing document, that
+				// doesn't have a leaseDoc.
+				t.logger.Warningf("missing lease document for existing document %q", doc.Id)
+				continue
+			}
+
+			if leaseDoc.Holder != doc.Holder {
+				// The holder for the document has changed. We shouldn't attempt
+				// to remove this document.
+				t.logger.Infof("lease expiry holder %q doesn't match the current lease holder %q, skipping", leaseDoc.Holder, doc.Holder)
```

Comment:
We should probably also include the leaseDoc.Id for context.
`"lease %q is currently held by %q but we were asked to expire it for %q, skipping"`